### PR TITLE
added movie update,create and delete hooks

### LINKED ISSUE #1330: Hooks to synchronize Algolia with Firestore movies
Hooks to synchronize Algolia with Firestore movies

### PR DESCRIPTION
**Pull request by Maxstgt**
close #1330 
movies are now synchronized with the algolia index

### CODE REVIEW

File: apps/backend-functions/src/movies.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  const after = change.after.data() as MovieDocument;
+
+  if (before === after) {
+    console.error('No changes detected');
+    throw new Error('movie update function got invalid movie data');
+  }
+
+  if (before.id === after.id) {
+    throw new Error('movie id can not be changed!');
+  }
```
Comment (ioaNikas) [inline apps/backend-functions/src/movies.ts]: You'll always throw here. You meant `before.id !== after.id` ?

File: apps/backend-functions/src/internals/algolia.ts [hunk 2]
```diff
@@ -42,3 +42,12 @@ export function storeSearchableMovie(
...
   return indexMoviesBuilder(adminKey).saveObject({ objectID: movie.id, movie });
 }
+
+export function deleteSearchableMovie(movieId: string): Promise<any> {
+  if (!algolia.adminKey) {
+    console.warn('No algolia id set, assuming dev config: skipping');
+    return Promise.resolve(true);
+  }
+
+  return indexMoviesBuilder().deleteObject(movieId);
```
Comment (lsenta) [inline apps/backend-functions/src/internals/algolia.ts]: Why is this line not using `adminKey` like line 43?

File: apps/backend-functions/src/movies.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+): Promise<any> {
+  const movie = snap.data();
+  const movieId = context.params.movieId;
+
+  if (!movie || !movieId) {
+    console.error('Invalid movie data:', movie);
+    throw new Error('movie update function got invalid movie data');
+  }
+
+  return Promise.all([
```
Comment (lsenta) [inline apps/backend-functions/src/movies.ts]: don't `Promise.all` just return the promise.

File: apps/backend-functions/src/movies.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    // Update algolia's index
+    storeSearchableMovie(movie, process.env['ALGOLIA_API_KEY'])
```
Comment (lsenta) [inline apps/backend-functions/src/movies.ts]: Can you see me a run of this code? As far as I know `process.env` does not work in google fuctions that's why we have functions.config.

Remove this, and look at `onOrganizationCreate`.

File: apps/backend-functions/src/movies.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  ]);
+}
+
+export function onMovieDelete(
+  // First param is needed? @Laurent
```
Comment (lsenta) [inline apps/backend-functions/src/movies.ts]: Don't `@people` in comments, we have tools for that.

Comment (lsenta) [reply apps/backend-functions/src/movies.ts]: https://firebase.google.com/docs/reference/functions/providers_firestore_.documentbuilder.html#on-delete

keep the argument.

File: apps/backend-functions/src/movies.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+export function onMovieUpdate(
+  change: functions.Change<FirebaseFirestore.DocumentSnapshot>,
+  context: functions.EventContext
+): Promise<any> {
+  const before = change.before.data() as MovieDocument;
+  const after = change.after.data() as MovieDocument;
+
+  if (before === after) {
```
Comment (lsenta) [inline apps/backend-functions/src/movies.ts]: This will NEVER happen, the only way for 2 objects to be equal is to point to the same reference.

I made you a jsfiddle:
https://jsfiddle.net/8gLvhy9q/2/

You're looking for Structural Equality: https://basarat.gitbooks.io/typescript/docs/javascript/equality.html

What you should do is compare the before and after fields that you send to algolia, if they are equal, skip the update. Else, send the update.

File: apps/backend-functions/src/main.ts [hunk 7]
```diff
@@ -96,7 +85,7 @@ export const admin = functions.https
 /** Trigger: when signature (`orgId`) is added to or removed from `validated[]`. */
 export const onDeliveryUpdateEvent = onDocumentUpdate('deliveries/{deliveryID}', onDeliveryUpdate);
 
-/** Trigger: when a stakeholder is added to a delivery. */
+/** Trigger: when a stakeholonOrganizationUpdateEventder is added to a delivery. */
```
Comment (lsenta) [inline apps/backend-functions/src/main.ts]: Undo this change

File: apps/backend-functions/src/main.ts [hunk 8]
```diff
@@ -42,18 +42,7 @@ export const onIpHashEvent = functions.pubsub.topic('eth-events.ipHash').onPubli
  *
  * NOTE: we will need to refactor the function to dispatch
  * depending on the file path. Or use different storage buckets
- * (one per app maybe).
- */
-export const generateHash = functions.storage
```
Comment (lsenta) [inline apps/backend-functions/src/main.ts]: that's out of scope for this PR, undo this change.

File: apps/backend-functions/src/main.ts [hunk 9]
```diff
@@ -42,18 +42,7 @@ export const onIpHashEvent = functions.pubsub.topic('eth-events.ipHash').onPubli
...
-  .onFinalize(hashToFirestore);
-
-/**
- * Trigger: when user creates an account.
- *
- * We create a corresponding document in `users/userID`.
- */
-export const onUserCreate = functions.auth
+ * (one per app maybe).onOrganizationUpdateEventfunctions.auth
   .user()
```
Comment (lsenta) [inline apps/backend-functions/src/main.ts]: Looks like a mistake, you commented out a lot of code. Do not keep dead code, remove it.

**Comment by lsenta** [comment]
You merge'd instead of rebasing, that's bad practice.

Have you tested all the code in there?

**Comment by Maxstgt** [comment]
yes, it is tested
